test/integration/apimachinery/admission_test.go

```diff
@@ -0,0 +1,342 @@
+	"k8s.io/kubernetes/test/integration/etcd"
+)
+
+var scheme = runtime.NewScheme()
+var codecs = serializer.NewCodecFactory(scheme)
+
+// Only add kinds to this list when this a virtual resource with get and create verbs that doesn't actually
+// store into it's kind.  We've used this downstream for mappings before.
+var kindWhiteList = sets.NewString()
+
+// namespace used for all tests, do not change this
+const testNamespace = "webhook-integration"
+
+type admissionElement struct {
+	gvk      metav1.GroupVersionKind
+	resource string
+	sub      string
+}
+
+type admissionTestData map[admissionElement]func() (*unstructured.Unstructured, error)
+
+// TestAdmission tests communication between API server and webhook process.
+func TestAdmission(t *testing.T) {
+	var lastGVR metav1.GroupVersionResource
+	var lastGVK metav1.GroupVersionKind
+	var lastOldGVR metav1.GroupVersionResource
+	var lastOldGVK metav1.GroupVersionKind
+	var lastOperation v1beta1.Operation
+
+	roots := x509.NewCertPool()
+	if !roots.AppendCertsFromPEM(localhostCert) {
+		t.Fatal("Failed to append Cert from PEM")
+	}
+	cert, err := tls.X509KeyPair(localhostCert, localhostKey)
+	if err != nil {
+		t.Fatalf("Failed to build cert with error: %+v", err)
+	}
+	webhook := &http.Server{
+		Addr: ":22443",
+		TLSConfig: &tls.Config{
+			RootCAs:      roots,
+			Certificates: []tls.Certificate{cert},
+		},
+	}
+	http.HandleFunc("/admission-endpoint", func(w http.ResponseWriter, r *http.Request) {
+		var body []byte
+		if r.Body != nil {
+			if data, err := ioutil.ReadAll(r.Body); err == nil {
+				body = data
+			}
+		}
+
+		// verify the content type is accurate
+		contentType := r.Header.Get("Content-Type")
+		if contentType != "application/json" {
+			t.Errorf("contentType=%s, expect application/json", contentType)
+		}
+		// The AdmissionReview that was sent to the webhook
+		requestedAdmissionReview := v1beta1.AdmissionReview{}
+
+		lastOperation = ""
+		lastGVR = metav1.GroupVersionResource{}
+		lastGVK = metav1.GroupVersionKind{}
+		lastOldGVR = metav1.GroupVersionResource{}
+		lastOldGVK = metav1.GroupVersionKind{}
+
+		// The AdmissionReview that will be returned
+		responseAdmissionReview := v1beta1.AdmissionReview{}
+		deserializer := codecs.UniversalDeserializer()
+		if _, _, err := deserializer.Decode(body, nil, &requestedAdmissionReview); err != nil {
+			t.Errorf("Fail to deserialize object: %s with error: %+v", string(body), err)
+			responseAdmissionReview.Response = response(false, err.Error())
+		} else {
+			responseAdmissionReview.Response = response(true, "Admitted by integration testing.")
+		}
+		t.Logf("Webhook operation: %s", requestedAdmissionReview.Request.Operation)
+		// Return the same UID
+		responseAdmissionReview.Response.UID = requestedAdmissionReview.Request.UID
+
+		lastOperation = requestedAdmissionReview.Request.Operation
+		lastGVR = requestedAdmissionReview.Request.Resource
+		lastGVK = requestedAdmissionReview.Request.Kind
```

Comment:
checking the attributes here is good. we also want to record the GVK of requestedAdmissionReview.Request.Object (if non-nil) and requestedAdmissionReview.Request.OldObject (if non-nil) and compare that to what we expect